[multi-chain] Implement a generic chain spec

Xanewok opened the pull request:
Closes #635 

For now this does not error out on unknown transaction types and interprets it as EIP-155 for the transactions and EIP-568 for the receipts, doing what we do on main. The integration test listed in #635 passes. What's left is not erroring out on missing nonce/mixHash for remote blocks. EDIT: That's now done in https://github.com/NomicFoundation/edr/pull/653/commits/d95eb9ab2be09d4ddfaf6e9d9214d17a73f08d6b.

@Wodann this is up for preliminary review if you feel like it, I tried to clean it up more or less and retain the existing structure found in edr_optimism.

Xanewok commented:
Ah, https://github.com/NomicFoundation/edr/pull/653/commits/d95eb9ab2be09d4ddfaf6e9d9214d17a73f08d6b regressed the `unknown_transaction_types`. Reverting to draft while I figure this out.

Wodann reviewed (commented):
Overall, this looks good! Thank you for the changes. The feedback on the chain spec trait design was also very helpful!

A couple of small comments/suggestions.

Once you have integrated this into Hardhat and validated that the Hardhat test suite passes with the generic chain type, this is good to merge - as far as I am concerned.

Xanewok commented on crates/edr_generic/src/lib.rs:
```diff
@@ -0,0 +1,18 @@
+/// [`L1ChainSpec`](edr_eth::chain_spec::L1ChainSpec).
+///
+/// Specifically:
+/// - it allows unknown transaction types (treates them as legacy
```
I swear I stared at this twice but couldn't quite figure out what's wrong - now I know what! Thanks :smile:

Xanewok commented on crates/edr_generic/src/rpc/block.rs:
```diff
@@ -0,0 +1,257 @@
+    /// Missing miner
+    #[error("Missing miner")]
+    MissingMiner,
+    #[error("Missing nonce")]
+    MissingNonce,
```
Ah, good catch! That's a leftover when I tested if the logic is correctly executed in the test.

Xanewok commented on crates/edr_generic/src/transaction.rs:
```diff
@@ -0,0 +1,190 @@
+// types different.
+#[repr(transparent)]
+#[derive(Clone, Debug, PartialEq, Eq, alloy_rlp::RlpEncodable)]
+pub struct SignedFallbackToPostEip155(pub transaction::Signed);
```
Sure, done in https://github.com/NomicFoundation/edr/pull/653/commits/80be89e822217ebf99b3d9e0e26fa5eadc2e1e01. Thanks!

Xanewok commented:
I have implemented the `ProviderSpec` and `SyncNapiSpec` and re-use the GenericChainSpec instead of the L1ChainSpec in edr_napi and updated the Hardhat patch to use it - let's see what CI says.

Xanewok commented on crates/edr_generic/src/eip2718.rs:
```diff
@@ -0,0 +1,145 @@
+            TypedEnvelope::Eip2930(_) => transaction::Type::Eip2930,
+            TypedEnvelope::Eip1559(_) => transaction::Type::Eip1559,
+            TypedEnvelope::Eip4844(_) => transaction::Type::Eip4844,
+            // TODO: Should we properly decode the transaction type?
+            TypedEnvelope::Unrecognized(_) => transaction::Type::Unrecognized(0xFF),
```
So I'm torn here, as in practice it shouldn't matter but for example you use the library to RLP-decode the typed envelope and saturate to `transaction::Type::Legacy` here, it would not be roundtrippable per se; maybe that's okay.

Do you want me to return `transaction::Type::Legacy` with a note or `unreachable!` or something else here, specifically?

Wodann replied:
That's correct. Internally, we never actually decoded anything that we didn't encode ourselves. That's why the guarantee holds in real use cases.

So the tricky case here is the `ChainSpecT::PooledTransaction::decode` in `pub fn handle_send_raw_transaction_request`. This should return an invalid transaction type error for any non-supported transaction types. As long as you return that for `TypedEnvelope::Unrecognized`, we should not be changing behaviour inadvertently.

Xanewok replied:
Hm, that is tricky - the `ChainSpecT::PooledTransaction` uses its decode impl but it's only compared against a `edr_eth::rlp::Error::Custom(&'static str)` and checks the string contents using a `edr_eth::transaction::Signed::is_invalid_transaction_type_error`.

While the `ProviderSpec` impl for `GenericChainSpec` has the PooledTransaction the same as edr_eth's, so we're safe, the interaction seems somewhat flimsy as we're crossing from the ChainSpec abstraction to the concrete edr_eth impls and use string checks, no less.

I'm also wary of introducing yet another type param but maybe, since we're already using generic and type-safe impls, we should introduce something like `TransactionDecodeError` and add our `InvalidType` variant on top of the RLP error or something like that? We seem to be recovering/handling this exact case often enough to uplift it to the type system but I'm curious to see what you think.

I took the liberty of matching on the const str value rather than using the if guard on the pattern in https://github.com/NomicFoundation/edr/pull/653/commits/31f123e238595cb953024f7948f88211a2abd332 for consistency - I hope you don't mind.

Xanewok commented on crates/edr_generic/src/rpc/transaction.rs:
```diff
@@ -0,0 +1,113 @@
+
+use crate::GenericChainSpec;
+
+// We need to use a newtype here as `RpcTypeFrom` cannot be implemented here,
```
In general there were few cases where it was not possible to re-use the original traits as we did depend on the actual trait implementation or the associated type was set for a given impl, rather than be generic (and so pluggable by potential other impls, like `(Fake)Sign`).

How often that's the case I'm not sure, as this was a very specific case of "do everything that L1 does except these few tiny bits" but I'd assume that we should provide sensible defaults or building blocks that the other implementers can more easily re-use what we have rather than copy the impls in some cases.

What's the ergonomic solution I'm not yet sure and I'm mostly just thinking out loud. Maybe allow for more generic impls and use the final marker trait impls that tie all of these together to validate that the intermediate types are compatible and expected?

For instance, I found it hard to wrap my head around what to implement as often the traits felt like they were circular, like in the case of satisfying both the `RpcSpec` and `ChainSpec` for a type. Another case is that some types like `EthBlockData` only use a single associated type from the `ChainSpec` and the `ChainSpec` feels like this tie-all/validation trait (which makes sense) but it made me harder to understand what's actually used and why. I had trouble deciphering some of the compiler errors and figuring out which trait impls are required because it felt so intertwined.

Maybe I just missed a proper mental model but I have to admit that it took me longer than it should :see_no_evil:

Xanewok commented:
@Wodann I hope everything looks good now! The tests are passing and I applied the patch; let me know if I should add or change anything. Thanks!

Wodann commented:
While going through the changes I noticed that there was a failing test in the EDR TS bindings but it wasn't being triggered in the CI.

I added CI support, causing [the failure](https://github.com/NomicFoundation/edr/actions/runs/10839293375/job/30079437545?pr=653), and pushed a fix for it.

I also re-added the implementation of the `NapiSpec` for `L1ChainSpec` as that should keep existing for future releases.

Wodann commented on crates/edr_napi/src/spec.rs:
```diff
@@ -85,6 +86,39 @@ pub trait SyncNapiSpec:
+    ) -> napi::Result<Response> {
+        let response = jsonrpc::ResponseData::from(response.map(|response| response.result));
+
+        serde_json::to_string(&response)
```
This code duplication will be removed in https://github.com/NomicFoundation/edr/pull/659/files#diff-bd82f1c79ac0cc9d088b3262a78277433ed1667e6f519c2ac84215247a6fcf61

Wodann reviewed (approved):
I made some changes to fix a failing test that wasn't running in CI. Other than that it LGTM!

There is one open discussion thread regarding a TODO. If my understanding is correct, feel free to proceed with merging, if you're happy with the changes that I made.

Xanewok commented:
@Wodann thanks for the fixes! For the one last thread I can confirm that it's working as you'd expect but left a note there as well but we can tackle this in a follow-up, if at all. Since the change in https://github.com/NomicFoundation/edr/pull/653/commits/31f123e238595cb953024f7948f88211a2abd332 is cosmetic, I'll go ahead and merge this to unblock your remaining work.
